aspnetcore/mvc/advanced/custom-model-binding.md

```diff
@@ -0,0 +1,108 @@
+
+The following example demonstrates how to accept and save a `byte[]` array that uses the `ByteArrayModelBinder`:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/ImageController.cs?name=post1&highlight=3)]
+
+You can POST a base64-encoded string to this api method using a tool like [Postman](https://www.getpostman.com/):
+
+![postman](custom-model-binding/images/postman.png "postman")
+
+The model binder will work equally well with a viewmodel instead of individual action arguments:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/ImageController.cs?name=post2&highlight=2)]
+
+As long as the binder can bind request data to appropriately named properties or arguments, model binding will succeed.
+
+## Implementing a custom model binder
+
+You can implement your own custom model binder that:
+
+- Converts incoming request data into properly-typed keys.
+- Uses Entity Framework Core to fetch the associated entity.
+- Passes the associated entity as an argument to the action method.
+
+The simplest approach is to use the `ModelBinderAttribute`. With this attribute, you specify the name of the input that is bound and the type of the `IModelBinderProvider` that it should use. The `ModelBinderAttribute` is applied to the type that should use the model binder:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Data/Author.cs?highlight=10)]
+
+The `AuthorEntityBinder` takes in an instance of `AppDbContext`. After confirming the expected input ("authorId" by default, or set using another `ModelBinderAttribute`) is present, it casts it to an integer and uses it to retrieve the associated entity. If successful, the result is set on the binding context. If model binding fails due to the input being the wrong type, a model state error is added. Otherwise, if no record is found for the id specified, the argument passed to the action method is set to `null`.
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Binders/AuthorEntityBinder.cs?name=demo)]
+
+The `AuthorEntityBinder` allows action methods to use the `Author` entity directly:
+
+[!code-csharp[Main](custom-model-binding/sample/CustomModelBindingSample/Controllers/BoundAuthorsController.cs?name=demo1&highlight=2)]
+
+In this example, since the name of the argument is not the default `authorId`, it's specified on the parameter using `ModelBinderAttribute`. Note that both the controller and action method are simplified compared to looking up the entity in the action method, since the `AppDbContext` is no longer required by the action or controller.
```

Comment:
Note that both the controller and action method are simplified compared to looking up the entity in the action method. The logic to fetch the author for EF is moved to the model binder. This can be considerable simplification when you have several methods that bind to the author model. (also DRY - right?)

 `Editor note, who cares about AppDbContext, that's trivial complexity``